### PR TITLE
[WIP][quantization] Re-write of Tensor-scalar add

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25622 [quantization] Re-write of Tensor-scalar add**
* #25545 [quantization] Dynamic dispatch for optimized quantized op kernels

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/qadd.cpp [hunk 1]
```diff
@@ -43,23 +43,59 @@ Tensor _add_out(Tensor& out, const Tensor& self, const Tensor& other) {
...
+
+  AT_DISPATCH_QINT_TYPES(self.scalar_type(), "qadd_scalar", [&]() {
+    double s = self.q_scale();
+    int64_t z = self.q_zero_point();
+    float c = other.toFloat();
+    int64_t q_min = std::numeric_limits<underlying_t>::min();
+    int64_t q_max = std::numeric_limits<underlying_t>::max();
+
+    int64_t rounded = std::nearbyint(c / s);
+    if (rounded > q_max || rounded < q_min) {
```
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/cpu/qadd.cpp]: This can be out of range, in these cases, we adjust the the scale, please remove lines 78 to 80

Comment (jamesr66a) [reply aten/src/ATen/native/quantized/cpu/qadd.cpp]: It doesn't work numerically under the automated testing. We can look at it tomorrow morning

File: aten/src/ATen/native/quantized/cpu/qadd.cpp [hunk 2]
```diff
@@ -94,16 +130,15 @@ class QAddOut final : public c10::OperatorKernel {
 template <bool ReLUFused = false>
 class QAddScalar final : public c10::OperatorKernel {
  public:
-  Tensor operator()(Tensor qa, Scalar b,
-                    double scale, int64_t zero_point) {
+  Tensor operator()(Tensor qa, Scalar b) {
```
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/cpu/qadd.cpp]: NIT: Typo in line 135

File: test/test_quantized.py [hunk 3]
```diff
@@ -147,25 +147,30 @@ def test_qadd_scalar_relu(self, A, b):
         add_scalar_relu = torch.ops.quantized.add_scalar_relu
 
         A, (scale, zero_point, dtype) = A
+        check_val = zero_point + b
+        if dtype == torch.qint8 and check_val < -128 or check_val > 127:
```
Comment (raghuramank100) [inline test/test_quantized.py]: These are corner cases that we should test for. , please remove

File: test/test_quantized.py [hunk 4]
```diff
@@ -147,25 +147,30 @@ def test_qadd_scalar_relu(self, A, b):
...
+            return
+        if dtype == torch.quint8 and check_val > 255 or check_val < 0:
+            return
+        if dtype == torch.quint8 and check_val > 2**31 - 1 or check_val < -(2**31):
```
Comment (raghuramank100) [inline test/test_quantized.py]: L 155: We should not support scalar_add for qint32.  Currently, qint32 support is restricted to zero-point = 0.

File: test/test_quantized.py [hunk 5]
```diff
@@ -147,25 +147,30 @@ def test_qadd_scalar_relu(self, A, b):
...
-        C_ref = torch.quantize_linear(C, scale, zero_point, dtype)
-        C_relu_ref = torch.quantize_linear(C_relu, scale, zero_point, dtype)
+        C_hat = add_scalar(qA, b)
+        C_relu_hat = add_scalar_relu(qA, b)
 
-        C_hat = add_scalar(qA, b, scale=scale, zero_point=zero_point)
-        C_relu_hat = add_scalar_relu(qA, b, scale=scale, zero_point=zero_point)
-
-        self.assertEqual(C_ref, C_hat,
+        self.assertEqual(C, C_hat.dequantize(), prec=1.0,
```
Comment (raghuramank100) [inline test/test_quantized.py]: Can we tighten the prec value? Does this still need to be 1?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please see comments, thanks

File: test/test_quantized.py (hunk 5 above)
Comment (jamesr66a) [reply test/test_quantized.py]: This fails for any arbitrary precision value

File: aten/src/ATen/native/quantized/cpu/qadd.cpp [hunk 6]
```diff
@@ -43,23 +43,59 @@ Tensor _add_out(Tensor& out, const Tensor& self, const Tensor& other) {
...
+  //    s' = [(q_max - z)*s + c]/[q_max - q_min]
+  //    z' = q_min
+  //
+  //  If q_max < z - round(c / s)
+  //    s' = -[(q_min - z)*s + c]/[q_max - q_min]
+  //    z' = q_max
+  //
+  //  Else
+  //    s' = s
+  //    z = z - round(c / s)
```
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/cpu/qadd.cpp]: The part that is missing is requantizing  the quantized values when the addition is out of bound:
We need to do:
 // Let s = scale, z = zero point, c = other.toFloat(),    c_q = round(c/s)
  //     q_min = lowest representable value of scalar type
  //     q_max = highest representable value of scalar type
  //
  //  Let s' = the calculated scale or the output
  //      z' = the calculated zero-point for the output
  If q_min > c_q
  //    s' = [(q_max - (z -  c_q)]/[q_max - q_min] *s
  //    z' = q_min
  //    Xq'  = torch.quantize_linear(Xq.dequantize() + c_q.dequantize() , s', z')
  //  If q_max < z - c_q
  //    s' = [z - c_q -q_min]/[q_max - q_min] *s
  //    z' = q_max
   //    Xq'  = torch.quantize_linear(Xq.dequantize() + c_q.dequantize(), s', z')

  //  Else
  //    s' = s
  //    z' = z - c_q

This op is identical in floating point semantics to:

 y = quantize( ( x_q.dequantize()+c_q.dequantize(), s',z') ).dequantize()

**Pull request closed without merging**